### PR TITLE
Github Automation App to monitor the critical workflows

### PR DESCRIPTION
**Pull request by prudhvigodithi**
### Description
Github Automation App to monitor the critical workflows:

- Creates the Alarms for each critical workflow.
- Creates the Alarms for each critical workflow with restrictive metric and namespace.
    - The app logic during runtime and based on the workflow result will update the workflow specific alarm metrics, related PR https://github.com/opensearch-project/automation-app/pull/21.
- The GitHub App role will have restrictive access to publish the alarm metrics.

### Issues Resolved
Part of https://github.com/opensearch-project/opensearch-build/issues/4941

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: DEVELOPER_GUIDE.md [hunk 1]
```diff
@@ -22,6 +22,8 @@ So you want to contribute code to this project? Excellent! We're glad you're her
     - `cdk deploy OpenSearchMetricsNginxReadonly`: To deploy the dashboard read only setup.
     - `cdk deploy OpenSearchWAF`: To deploy the AWS WAF for the project ALB's.
     - `cdk deploy OpenSearchMetrics-Monitoring`: To deploy the alerting stack which will monitor the step functions and URL of the project coming from [METRICS_HOSTED_ZONE](https://github.com/opensearch-project/opensearch-metrics/blob/main/infrastructure/lib/enums/project.ts)
+    - `cdk deploy OpenSearchMetrics-GitHubAutomationApp-Secret`: Creates the GitHub app secret which will be used during the GitHub app runtime.
```
Comment (peterzhuamazon) [inline DEVELOPER_GUIDE.md:25]: I thought we already have the secrets created?

File: infrastructure/lib/infrastructure-stack.ts [hunk 2]
```diff
@@ -20,6 +20,7 @@ import { OpenSearchMetricsNginxCognito } from "./constructs/opensearchNginxProxy
 import { OpenSearchMetricsMonitoringStack } from "./stacks/monitoringDashboard";
 import { OpenSearchMetricsSecretsStack } from "./stacks/secrets";
 import { GitHubAutomationApp } from "./stacks/gitHubAutomationApp";
+import {GitHubWorkflowMonitorAlarms} from "./stacks/gitHubWorkflowMonitorAlarms";
```
Comment (peterzhuamazon) [inline infrastructure/lib/infrastructure-stack.ts:23]: Need formating.

**Review by bshien — APPROVED** [review]

File: infrastructure/lib/stacks/gitHubWorkflowMonitorAlarms.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+export interface AlarmProps {
+    readonly namespace: string;
+    readonly metricName: string;
+    readonly workflows: string[];
+}
+
+export class GitHubWorkflowMonitorAlarms extends Stack {
+    readonly workflowAlarmsArn: string[] = [];
+    readonly metricName: string
```
Comment (rishabh6788) [inline infrastructure/lib/stacks/gitHubWorkflowMonitorAlarms.ts:21]: is this being used anywhere? I only see `props.metricName` being used, the other one is Metrics class property.

File: infrastructure/lib/stacks/gitHubWorkflowMonitorAlarms.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    constructor(scope: Construct, id: string, props: AlarmProps) {
+        super(scope, id);
+        props.workflows.forEach(workflow => {
+            const dimensionValue = workflow;
+            const workflowMetric = new Metric({
+                namespace: props.namespace,
+                metricName: props.metricName,
```
Comment (rishabh6788) [inline infrastructure/lib/stacks/gitHubWorkflowMonitorAlarms.ts:28]: I assume this metric would already be flowing into CW through the general workflow, right?

Comment (rishabh6788) [reply infrastructure/lib/stacks/gitHubWorkflowMonitorAlarms.ts:28]: Ignore, this is just creating a metric and data will flow later once this metric object is created.

Comment (prudhvigodithi) [reply infrastructure/lib/stacks/gitHubWorkflowMonitorAlarms.ts:28]: That is correct Rishab, coming from this PR https://github.com/opensearch-project/automation-app/pull/21 the data will flow from the app events.

File: infrastructure/lib/stacks/gitHubWorkflowMonitorAlarms.ts (hunk 3 above)
Comment (prudhvigodithi) [reply infrastructure/lib/stacks/gitHubWorkflowMonitorAlarms.ts:21]: Ya this is used to create the `GitHubWorkflowMonitorAlarms` CDK stack, coming from `infrastructure/lib/infrastructure-stack.ts`

File: DEVELOPER_GUIDE.md (hunk 1 above)
Comment (prudhvigodithi) [reply DEVELOPER_GUIDE.md:25]: Ya this is created Peter but adding it to the DEVELOPER_GUIDE.md, once we call the `OpenSearchMetrics-GitHubAutomationApp` stack the `OpenSearchMetrics-GitHubAutomationApp-Secret` is created as dependency. Also we can call the `OpenSearchMetrics-GitHubAutomationApp-Secret` separately hence recording this to the DEVELOPER_GUIDE.md.

File: infrastructure/lib/infrastructure-stack.ts (hunk 2 above)
Comment (prudhvigodithi) [reply infrastructure/lib/infrastructure-stack.ts]: Updated peter.

**Review by rishabh6788 — APPROVED** [review]

**Review by peterzhuamazon — APPROVED** [review]

**Comment by prudhvigodithi** [comment]
Thanks Peter, Rishab and Brandon for the review, since there are known issues with `Mend Security Check` I will merge this PR.

**Pull request merged**